[#58][Thread] feature : 쓰레드 레디스 구현

leedongha1998 opened the pull request:
## Description
- 쓰레드 생성
- 쓰레드 댓글 여론 분석
- 통합 테스트
- 댓글 client 구현

Resolves: #58 

## PR Type
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Commit:
[#58][Thread] feature : 쓰레드 레디스 구현
- 쓰레드 생성
- 쓰레드 댓글 여론 분석
- 통합 테스트
- 댓글 client 구현

Commit:
[#58][Thread] refactor : 댓글 client url 수정

Commit:
[#58][Thread] refactor : Thread 레디스 설정 변경

venus-y commented on thread/src/main/java/com/newpick4u/thread/thread/application/usecase/ThreadServiceImpl.java:
```diff
@@ -1,50 +1,239 @@
+        // 쓰레드에 달린 댓글들 가져오기
+        List<CommentResponse> comments = commentClient.getAllByThreadId(UUID.fromString(threadId));
+        if (comments == null || comments.isEmpty()) {
+          throw new IllegalArgumentException("댓글이 없습니다.");
```
`CollectionUtils.isEmpty()`로 코드를 조금 더 간소화 할 수 있을 것 같습니다!

hp5234 replied:
다수의 쓰레드들에 대한 작업이 하나의 트랜잭션으로 처리되어야할 필요가 있는지, 그리고 각각 댓글서비스조회 -> ai 요청 -> 쓰레드 조회 -> 쓰레드 업데이트 로직이 하나의 트랜잭션으로 처리되어야할 필요가 있는지 고민해볼수 있을거같습니다.

leedongha1998 replied:
수정 완료했습니다

venus-y commented on thread/src/main/java/com/newpick4u/thread/thread/application/usecase/ThreadServiceImpl.java:
```diff
@@ -1,50 +1,239 @@
+    Set<String> processedTags = updateExistingThreads(hotTags, tagToThreadMap);
+
+    int threadsToCreate = MAX_THREADS - openThreads.size();
+
```
최대 쓰레드 개수에서 오픈쓰레드의 사이즈만큼 빼는 부분은 메서드로 추출해주셔도 좋을 것 같습니다!

leedongha1998 replied:
수정 완료했습니다

venus-y commented on thread/src/main/java/com/newpick4u/thread/thread/application/usecase/ThreadServiceImpl.java:
```diff
@@ -1,50 +1,239 @@
+  private Set<String> getHotTagsFromRedis() {
+    Set<ZSetOperations.TypedTuple<String>> sortedTags = redisTemplate.opsForZSet()
+        .reverseRangeByScoreWithScores(HOT_TAG_KEY, 30, Double.MAX_VALUE);
+
```
30도 상수로 관리해주셔도 좋을 것 같습니다.

leedongha1998 replied:
수정 완료했습니다

venus-y reviewed (approved):
고생하셨습니다:)

Commit:
[#58][Thread] refactor : ThreadServiceImpl 수정
- 새로운 쓰레드 추가 부분 메소드로 추출
- Collection 빈 값 체크 CollectionUtils.isEmpty()로 변경
- 태그 최소 점수 상수로 변경

Commit:
[#58][Thread] refactor : ThreadServiceImpl 수정

hp5234 commented on thread/build.gradle:
```diff
@@ -30,11 +30,13 @@ repositories {
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.kafka:spring-kafka'
-    implementation 'org.springframework.cloud:spring-cloud-starter-netflix-eureka-client'
+//    implementation 'org.springframework.cloud:spring-cloud-starter-netflix-eureka-client'
```
유레카 의존성이 필요하지 않을까요?

hp5234 commented on thread/src/main/java/com/newpick4u/thread/thread/domain/entity/Thread.java:
```diff
@@ -22,17 +27,36 @@ public class Thread extends BaseEntity {
-  @Column(name = "news_id")
-  private UUID newsId;
+  @Column(name = "tag_name", unique = true, nullable = false)
+  private String tagName;
```
태그 이름이 유니크면 한번 열렸던 태그의 쓰레드가 재오픈 되는일은 없을까 고민해볼 수 있을것같습니다. 
ex : 삼성전자 쓰레드 오픈 -> close -> 삼성전자 오픈

hp5234 reviewed (approved):
고생하셨습니다.
